Marlin/src/lcd/extui/dgus_e3s1pro/DGUSDisplay.h

```diff
@@ -0,0 +1,185 @@
+  // Play sounds using the display speaker.
+  //   start: position at which the sound was stored on the display.
+  //   len: how many sounds to play. Sounds will play consecutively from start to start+len-1.
+  //   volume: playback volume. 0 keeps the current volume.
+  static void PlaySound(uint8_t start, uint8_t len=1, uint8_t volume=0);
+  // Enable/disable a specific touch control.
+  //   type: control type.
+  //   control: index of the control on the page (set during screen development).
+  static void EnableControl(DGUS_Screen screen, DGUS_ControlType type, DGUS_Control control);
+  static void DisableControl(DGUS_Screen screen, DGUS_ControlType type, DGUS_Control control);
+
+  static uint8_t GetBrightness();
+  static uint8_t GetVolume();
+
+  // Set the display brightness/volume, ranging 0 - 100
+  static void SetBrightness(uint8_t brightness);
+  static void SetVolume(uint8_t volume);
+
+  // Periodic tasks, eg. Rx-Queue handling.
+  static void Loop();
+
+  // Helper for users of this class to estimate if an interaction would be blocking.
+  static size_t GetFreeTxBuffer();
+  static void FlushTx();
+
+  // Checks two things: Can we confirm the presence of the display and has we initialized it.
+  // (both boils down that the display answered to our chatting)
+  static bool IsInitialized() {
+    return initialized;
+  }
+
+  static uint8_t gui_version;
+  static uint8_t os_version;
+
+  template<typename T>
+  static T SwapBytes(const T value) {
+    union {
+      T val;
+      char byte[sizeof(T)];
+    } src, dst;
+
+    src.val = value;
+    LOOP_L_N(i, sizeof(T)) dst.byte[i] = src.byte[sizeof(T) - i - 1];
+    return dst.val;
+  }
+
+  template<typename T_in, typename T_out, uint8_t decimals>
+  T_out FromFixedPoint(const T_in value) {
+    return (T_out)((float)value / POW(10, decimals));
+  }
+
+  template<typename T_in, typename T_out, uint8_t decimals>
+  T_out ToFixedPoint(const T_in value) {
+    return (T_out)LROUND((float)value * POW(10, decimals));
+  }
+
+private:
+  enum dgus_header : uint8_t {
+    DGUS_HEADER1 = 0x5A,
+    DGUS_HEADER2 = 0xA5
+  };
+
+  enum dgus_command : uint8_t {
+    DGUS_WRITEVAR = 0x82,
+    DGUS_READVAR = 0x83
+  };
+
+  enum rx_datagram_state_t : uint8_t {
+    DGUS_IDLE,           //< waiting for DGUS_HEADER1.
+    DGUS_HEADER1_SEEN,   //< DGUS_HEADER1 received
+    DGUS_HEADER2_SEEN,   //< DGUS_HEADER2 received
+    DGUS_WAIT_TELEGRAM,  //< LEN received, Waiting for to receive all bytes.
+  };
+
+  enum dgus_system_addr : uint16_t {
+    DGUS_VERSION = 0x000F // OS/GUI version
+  };
+
+  static void WriteHeader(uint16_t addr, uint8_t command, uint8_t len);
+  static void ProcessRx();
+
+  static uint8_t volume;
+  static uint8_t brightness;
+
+  static rx_datagram_state_t rx_datagram_state;
+  static uint8_t rx_datagram_len;
+
+  static bool initialized;
+};
+
+template<> inline uint16_t DGUSDisplay::SwapBytes(const uint16_t value) {
+  return ((value << 8) | (value >> 8));
```

Comment:
My guess is that this needs to be made contingent on the endianness of the target. We have no guarantee that all targets for this display will have the same endianness.

Reply:
Looks like we have a nice simlpe endian header now.

Reply:
Note that although it should work fine on most/all platforms it's not strictly speaking C++ compliant since type punning is undefined behavior. I have yet to see a compiler getting angry at that though.

The generic swap() template I took from the DGUS reloaded code yields the same assembly as GCC's builtins __builtin_bswap16/32/64 so I haven't looked into optimizing it further.